### PR TITLE
[Tailcall] Better sizing of tailcalls, allow arm64 r4, copy arm64 stacked parameters

### PR DESCRIPTION
**Pull request by jaykrell**
Better sizing of tailcalls via new OP_TAILCALL_PARAMETER.
Allow arm64 r4.
Copy arm64 stacked parameters.
Better interface to instruction sizes.
Adding possibly missing reallocs.

### CODE REVIEW

File: mono/mini/mini-arm.c [hunk 1]
```diff
@@ -1758,7 +1758,10 @@ mono_arch_tailcall_supported (MonoCompile *cfg, MonoMethodSignature *caller_sig,
 	if (!res && !debug_tailcall)
 		goto exit;
 
-	res &= IS_SUPPORTED_TAILCALL (callee_info->stack_usage <= 16 * 4); // FIXME why?
+	// FIXME The limit here is that moving the parameters requires addressing the parameters
+	// with 12bit (4K) immediate offsets.
+	res &= IS_SUPPORTED_TAILCALL (callee_info->stack_usage < 4096);
```
Comment (jaykrell) [inline mono/mini/mini-arm.c]: Maybe feels safer to stay away from the limit? 1k? 2k, 3k?

**Comment by jaykrell** [comment]
Thoughts?

File: mono/mini/mini.h [hunk 2]
```diff
@@ -1698,6 +1698,12 @@ extern const char MONO_ARCH_CPU_SPEC [];
 extern const guint16 MONO_ARCH_CPU_SPEC_IDX(MONO_ARCH_CPU_SPEC) [];
 #define ins_get_spec(op) ((const char*)&MONO_ARCH_CPU_SPEC + MONO_ARCH_CPU_SPEC_IDX(MONO_ARCH_CPU_SPEC)[(op) - OP_LOAD])
 
+static inline unsigned
+ins_get_size (int opcode)
```
Comment (vargaz) [inline mono/mini/mini.h]: This can return simply an int.

File: mono/mini/mini-arm64.c [hunk 3]
```diff
@@ -4340,7 +4340,13 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
...
 			}
 
-			// FIXME: Copy stack arguments
+			// Copy stack arguments.
+			// FIXME a fixed size memcpy is desirable here,
+			// at least for larger values of stack_usage.
+			for (i = 0; i < call->stack_usage; i += sizeof (mgreg_t)) {
+				code = emit_ldrx (code, ARMREG_LR, ARMREG_SP, i);
+				code = emit_strx (code, ARMREG_LR, ARMREG_R28, i);
+			}
```
Comment (vargaz) [inline mono/mini/mini-arm64.c]: Why does this use R28 ?

File: mono/mini/method-to-ir.c [hunk 4]
```diff
@@ -8408,10 +8428,13 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				/* Handle tailcalls similarly to calls */
 				DISABLE_AOT (cfg);
 
+				MonoMethodSignature *sig = mono_method_signature (cmethod);
+
+				emit_tailcall_parameters (cfg, sig);
```
Comment (vargaz) [inline mono/mini/method-to-ir.c]: There is already an fsig computed above.

File: mono/mini/method-to-ir.c [hunk 5]
```diff
@@ -8408,10 +8428,13 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
...
 				MONO_INST_NEW_CALL (cfg, call, OP_TAILCALL);
 				call->method = cmethod;
 				call->tailcall = TRUE;
-				call->signature = mono_method_signature (cmethod);
+				call->signature = sig;
 				call->args = (MonoInst **)mono_mempool_alloc (cfg->mempool, sizeof (MonoInst*) * n);
```
Comment (vargaz) [inline mono/mini/method-to-ir.c]: This should use fsig.

**Comment by vargaz** [comment]
Looks ok otherwise.

File: mono/mini/mini.h (hunk 2 above)
Comment (jaykrell) [reply mono/mini/mini.h]: Ok. It has to cast to unsigned char first though, to get the full 0-255 range.

File: mono/mini/mini-arm64.c (hunk 3 above)
Comment (jaykrell) [reply mono/mini/mini-arm64.c]: What should it use? cfg->arch.args_reg?

Comment (jaykrell) [reply mono/mini/mini-arm64.c]: Same thing, right?

**Comment by jaykrell** [comment]
I wonder if this shouldn't be two separate PRs.
They are related -- the sizing is needed, if we are to loop over stacked parameters without bound.
r4 and stacked arm64 parameter handling could be handled together.
At least the subject lists it all.

File: mono/mini/method-to-ir.c (hunk 5 above)
Comment (jaykrell) [reply mono/mini/method-to-ir.c]: Agreed, fixed.

File: mono/mini/method-to-ir.c (hunk 4 above)
Comment (jaykrell) [reply mono/mini/method-to-ir.c]: Agreed, fixed.

File: mono/mini/mini-arm64.c (hunk 3 above)
Comment (jaykrell) [reply mono/mini/mini-arm64.c]: I still confused on this one -- what should it be?

File: mono/mini/mini-amd64.c [hunk 6]
```diff
@@ -4650,12 +4661,9 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 
 			g_assert (!cfg->method->save_lmf);
 
-			/* the size of the tailcall op depends on signature, let's check for enough
-			 * space in the code buffer here again */
-			max_len += AMD64_NREG * 4 + call->stack_usage * 15 + EXTRA_CODE_SPACE;
-			max_len += 64; // FIXME make this two pass for an accurate size
-
-			if (G_UNLIKELY (offset + max_len > cfg->code_size)) {
+			max_len += AMD64_NREG * 4;
```
Comment (jaykrell) [inline mono/mini/mini-amd64.c]: This AMD64_NREG  * 4 should be part of the .md,and perhaps is, but leaving it alone for now.

File: mono/mini/mini-amd64.c [hunk 7]
```diff
@@ -4650,12 +4661,9 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
...
+			max_len += call->stack_usage / sizeof (mgreg_t) * ins_get_size (OP_TAILCALL_PARAMETER);
```
Comment (jaykrell) [inline mono/mini/mini-amd64.c]: Removing this line is kinda the point, depending on how else the rest of the code works, which I'm not sure of.

File: mono/mini/mini-arm64.c [hunk 8]
```diff
@@ -4331,14 +4351,23 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 				g_assert (ins->inst_basereg != -1);
 				g_assert (ins->inst_basereg != ARMREG_IP0);
 				g_assert (ins->inst_basereg != ARMREG_IP1);
+				g_assert (ins->inst_basereg != ARMREG_LR);
```
Comment (jaykrell) [inline mono/mini/mini-arm64.c]: assert could be relaxed, i.e. if stack_usage == 0, which is common, but ok.

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Comment by jaykrell** [comment]
I'm a little confused as to where accurate length is needed.
call updates max_len, e.g. arm emit_float_args.
For realloc purposes, length can be computed late.
Is that it? Or should arm call .md be increased?

**Pull request merged**